### PR TITLE
Http only WIP

### LINKED ISSUE #150: RFC: Change the default protocol to http
The current protocol is `node`. This means Logstash joins the cluster. The cons of this `node` default are:

* bidirectional communication must be possible between Logstash and the _entire_ cluster.
* Logstash's NodeClient can serve as a client node and answer search queries. Probably not expected.
* Shield users cannot use NodeClient, anyway, without granting Logstash full access to the cluster as a member of the cluster (full privileges)
* The default NodeClient uses multicast for discovery, and some operating systems (like CentOS, I think?) and networks (like AWS EC2's network) block this by default.
* Exceptions in Elasticsearch are serialized over the network using the standard Java Serializable. In our experience, Java's Serializable isn't stable across different JVM versions (exceptions will fail to deserialize).

Changing to HTTP as the default would make it:
* easier to integrate with Shield
* easier to debug (http is simpler than the transport protocol, probably)
* independent of Elasticsearch ABI (transport serialization)
* independent of JVM version
* Logstash release could be shipped _without_ the Elasticsearch libraries (saves 10-20mb)

Downsides:
* We'd want to have a default `host` setting, `localhost` is probably OK.
* We need discovery and sniffing enabled in the http protocol also.

### PR DESCRIPTION
**Pull request by andrewvc**
This is a WIP for efforts to remove support for non-HTTP protocols 

Fixes #150

### CODE REVIEW

File: spec/es_spec_helper.rb [hunk 1]
```diff
@@ -36,6 +31,7 @@ def get_client
 
   # this :all hook gets run before every describe block that is tagged with :integration => true.
   config.before(:all, :integration => true) do
+    require 'pry'; binding.pry
```
Comment (talevy) [inline spec/es_spec_helper.rb]: WIP //here just as a reminder

**Comment by talevy** [comment]
we can take a look at simplifying the retry logic, since we do not need to normalize the response across protocols.

**Comment by andrewvc** [comment]
@talevy what specific ideas did you have wrt cleaning up the retry code? Was this the Stud::Buffer cleanup you mentioned earlier we decided to forego?

BTW, I think this is getting close to good for an LGTM. I don't plan on merging this till we have logstash-output-elasticsearch_native ready

**Comment by talevy** [comment]
@andrewvc 

Nothing for this PR, but I will make a follow-up PR to make the information nicer. I have the "normalize" step for the error message returned from a bulk request. this normalization was done mainly to help wrangle between the response from the java client and the http client. Now that we only have one, this step is not necessary.

**Comment by andrewvc** [comment]
Mirrored the present content of this repo into https://github.com/logstash-plugins/logstash-output-elasticsearch_native to begin refactor there to remove HTTP.

**Comment by andrewvc** [comment]
@talevy @jordansissel what are your thoughts on the current state of this plugin? What do we need to get to LGTM?

File: lib/logstash/outputs/elasticsearch/protocol.rb [hunk 2]
```diff
@@ -1,307 +1,93 @@
...
-
-      def template_exists?(name)
-        @client.indices.get_template(:name => name)
-        return true
-      rescue Elasticsearch::Transport::Transport::Errors::NotFound
-        return false
-      end # def template_exists?
+    def bulk(actions)
+      bulk_response = @client.bulk(:body => actions.collect do |action, args, source|
+                                     if source
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch/protocol.rb]: large indentation intended?

**Comment by talevy** [comment]
@andrewvc running this plugin against elasticsearch-shield/secure tests should be a blocker. @jsvd, how can we run this against your elasticsearch-secure specs?

File: lib/logstash/outputs/elasticsearch/protocol.rb (hunk 2 above)
Comment (andrewvc) [inline lib/logstash/outputs/elasticsearch/protocol.rb]: ahhhh good catch.

File: lib/logstash/outputs/elasticsearch.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # encoding: utf-8
-require "logstash/namespace"
+require "logstash/pipeline"
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: why this change?

File: lib/logstash/outputs/elasticsearch.rb [hunk 4]
```diff
@@ -8,32 +8,20 @@
...
-# as configuration options, there are two methods:
-#
-# * Create an `elasticsearch.yml` file in the $PWD of the Logstash process
-# * Pass in es.* java properties (`java -Des.node.foo=` or `ruby -J-Des.node.foo=`)
-#
-# With the default `protocol` setting ("node"), this plugin will join your
-# Elasticsearch cluster as a client node, so it will show up in Elasticsearch's
-# cluster status.
+# We strongly encourage the use of HTTP over the node protocol. It is just as
+# fast and far easier to administer. For those wishing to use the java protocol please see the 'elasticsearch_java' gem.
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: we should probably provide a direct link here in the future.

File: lib/logstash/outputs/elasticsearch.rb (hunk 3 above)
Comment (andrewvc) [inline lib/logstash/outputs/elasticsearch.rb]: TBH I don't recall . I'll switch it back

File: lib/logstash/outputs/elasticsearch.rb (hunk 4 above)
Comment (andrewvc) [inline lib/logstash/outputs/elasticsearch.rb]: Yep, once we have a docs page up.

File: lib/logstash/outputs/elasticsearch.rb [hunk 5]
```diff
@@ -165,44 +122,9 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
...
   config :host, :validate => :array
 
-  # The port for Elasticsearch transport to use.
-  #
-  # If you do not set this, the following defaults are used:
-  # * `protocol => http` - port 9200
-  # * `protocol => transport` - port 9300-9305
-  # * `protocol => node` - port 9300-9305
+  # You can set the remote port as part of the host, or explicitly here as well
   config :port, :validate => :string
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: we can default to `9200` now, no?

File: lib/logstash/outputs/elasticsearch.rb [hunk 6]
```diff
@@ -257,6 +153,7 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
   # - index: indexes a document (an event from Logstash).
   # - delete: deletes a document by id
   # - create: indexes a document, fails if a document by that id already exists in the index.
+  # - update: updates a document by id
   # following action is not supported by HTTP protocol
   # - create_unless_exists: creates a document, fails if no id is provided
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: we can just remove this option from the the http-only repo entirely

File: lib/logstash/outputs/elasticsearch.rb [hunk 7]
```diff
@@ -329,101 +237,41 @@ def register
...
-        client_settings["node.name"] = @node_name
-      else
-        client_settings["node.name"] = "logstash-#{Socket.gethostname}-#{$$}-#{object_id}"
-      end
-
-      @@plugins.each do |plugin|
-        name = plugin.name.split('-')[-1]
-        client_settings.merge!(LogStash::Outputs::ElasticSearch.const_get(name.capitalize).create_client_config(self))
-      end
+    if @action == "create_unless_exists"
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: see note above, but I think we can go ahead and remove this since it does not apply to http.

File: lib/logstash/outputs/elasticsearch.rb [hunk 8]
```diff
@@ -329,101 +237,41 @@ def register
...
+    client_settings[:path] = "/#{@path}/".gsub(/\/+/, "/") # Normalize slashes
+    @logger.debug? && @logger.debug("Normalizing http path", :path => @path, :normalized => client_settings[:path])
 
-    if @port.nil?
-      @port = case @protocol
-        when "http"; "9200"
-        when "transport", "node"; "9300-9305"
-      end
-    end
+    @port ||= "9200"
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: I'd prefer having this type of logic via the `default` setting in the config declaration above

File: lib/logstash/outputs/elasticsearch.rb [hunk 9]
```diff
@@ -329,101 +237,41 @@ def register
...
 
-    if @host.nil? && @protocol != "node" # node can use zen discovery
+    if @host.nil?
       @logger.info("No 'host' set in elasticsearch output. Defaulting to localhost")
       @host = ["localhost"]
```
Comment (talevy) [inline lib/logstash/outputs/elasticsearch.rb]: does it even make sense to ever have host nil in the http-only world? we can just default this to localhost above and get rid of this logic. what do you think?

File: spec/es_spec_helper.rb [hunk 10]
```diff
@@ -9,57 +9,61 @@
 CONTAINER_IMAGE = "elasticsearch"
 CONTAINER_TAG = "1.6"
 
-module ESHelper
+DOCKER_INTEGRATION = ENV["DOCKER_INTEGRATION"]
```
Comment (talevy) [inline spec/es_spec_helper.rb]: should we also allow environment variables to explicitly target an ES instance with custom host/port?

this can wait for another follow-up PR if we think its a useful option

File: spec/es_spec_helper.rb [hunk 11]
```diff
@@ -9,57 +9,61 @@
...
+    config.before(:all, :integration => true) do
+
+
+      # check if container exists already before creating new one.
+      begin
+        ls = Longshoreman::new
+        ls.container.get(CONTAINER_NAME)
+      rescue Docker::Error::NotFoundError
+        Longshoreman.new("#{CONTAINER_IMAGE}:#{CONTAINER_TAG}", CONTAINER_NAME)
+        # TODO(talevy): verify ES is running instead of static timeout
```
Comment (talevy) [inline spec/es_spec_helper.rb]: making a note here for a possible follow-up PR.

maybe we can add an endpoint in the http_client wrapper we have going to check if ES is up. then we can loop-tick the check here until it says yes here as a replacement of the long-optimistic 10-second wait here

File: spec/es_spec_helper.rb (hunk 10 above)
Comment (andrewvc) [inline spec/es_spec_helper.rb]: I think it's not a great benefit at the moment. YAGNI IMHO

File: spec/unit/outputs/elasticsearch_spec.rb [hunk 12]
```diff
@@ -2,48 +2,18 @@
 
 describe "outputs/elasticsearch" do
   context "registration" do
-    it "should register" do
-      output = LogStash::Plugin.lookup("output", "elasticsearch").new("embedded" => "false", "protocol" => "transport", "manage_template" => "false")
-      # register will try to load jars and raise if it cannot find jars
-      expect {output.register}.to_not raise_error
-    end
-
     it "should fail to register when protocol => http, action => create_unless_exists" do
```
Comment (talevy) [inline spec/unit/outputs/elasticsearch_spec.rb]: if we decide to remove this option from the vocabulary of this plugin, we can remove this test as well. woohoo!

File: spec/unit/outputs/elasticsearch_ssl_spec.rb [hunk 13]
```diff
@@ -2,81 +2,56 @@
...
-          "ssl" => true,
-          "keystore" => keystore_path,
-          "keystore_password" => "test"
-        }
-        next LogStash::Outputs::ElasticSearch.new(settings)
-      end
+    subject do
+      require "logstash/outputs/elasticsearch"
+      settings = {
+        "host" => "node01",
```
Comment (talevy) [inline spec/unit/outputs/elasticsearch_ssl_spec.rb]: maybe now is a good time (another possible follow-up PR!) to work with @jsvd to finish up codifying this setup from within the plugin tests so that we can run these locally as well.

Comment (andrewvc) [inline spec/unit/outputs/elasticsearch_ssl_spec.rb]: I'm missing where/how this host is defined. Any context @talevy ?

File: spec/es_spec_helper.rb (hunk 10 above)
Comment (talevy) [inline spec/es_spec_helper.rb]: sounds good, I don't feel strongly about it

File: spec/unit/outputs/elasticsearch_ssl_spec.rb (hunk 13 above)
Comment (talevy) [inline spec/unit/outputs/elasticsearch_ssl_spec.rb]: thats just it, the context is not localized to this repo. it is a custom setup that is done for internal testing

**Comment by talevy** [comment]
LGTM

File: lib/logstash/outputs/elasticsearch.rb [hunk 14]
```diff
@@ -8,32 +8,20 @@
...
-require 'logstash-output-elasticsearch_jars.rb'
+require "logstash/outputs/elasticsearch/http_client"
 
 # This output lets you store logs in Elasticsearch and is the most recommended
 # output for Logstash. If you plan on using the Kibana web interface, you'll
-# need to use this output.
+# want to use this output.
 #
-# *VERSION NOTE*: Your Elasticsearch cluster must be running Elasticsearch 1.0.0 or later.
+# This output only speaks the HTTP, which is the preferred protocol for interacting with elasticsearch. By default
```
Comment (suyograo) [inline lib/logstash/outputs/elasticsearch.rb]: Can we capitalize elasticsearch?

Comment (andrewvc) [inline lib/logstash/outputs/elasticsearch.rb]: For sure! (/me remembers when in all official copy Elasticsearch was always lowercased)

Comment (suyograo) [inline lib/logstash/outputs/elasticsearch.rb]: sorry, nit pick :)

File: lib/logstash/outputs/elasticsearch.rb [hunk 15]
```diff
@@ -165,43 +122,8 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
 
   config :host, :validate => :array
```
Comment (suyograo) [inline lib/logstash/outputs/elasticsearch.rb]: @andrewvc what are your thoughts on deprecating `host` and introducing `hosts`. See https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/172

Comment (andrewvc) [inline lib/logstash/outputs/elasticsearch.rb]: We should definitely put this in before 2.0

File: lib/logstash/outputs/elasticsearch.rb [hunk 16]
```diff
@@ -165,43 +122,8 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
...
-  # This setting no longer does anything. It exists to keep config validation
-  # from failing. It will be removed in future versions.
-  config :max_inflight_requests, :validate => :number, :default => 50, :deprecated => true
-
-  # The node name Elasticsearch will use when joining a cluster.
-  #
-  # By default, this is generated internally by the ES client.
-  config :node_name, :validate => :string
+  # You can set the remote port as part of the host, or explicitly here as well
+  config :port, :validate => :string, :default => 9200
```
Comment (suyograo) [inline lib/logstash/outputs/elasticsearch.rb]: do we need to deprecate or remove this if we support `host:port`. Just an opportunity to clean up here..

File: spec/es_spec_helper.rb (hunk 11 above)
Comment (suyograo) [inline spec/es_spec_helper.rb]: @talevy I would create an issue for this. This will get lost otherwise

**Comment by andrewvc** [comment]
This has been merged in #240
